server/src/main/java/org/apache/druid/server/coordination/SegmentLoadDropHandler.java

```diff
@@ -789,15 +789,20 @@ private CustomSettableFuture(
 
     public void resolve()
     {
-      synchronized (statusRefs) {
+      synchronized (requestStatusesLock) {
         if (isDone()) {
           return;
         }
 
-        List<DataSegmentChangeRequestAndStatus> result = new ArrayList<>(statusRefs.size());
-        statusRefs.forEach(
-            (request, statusRef) -> result.add(new DataSegmentChangeRequestAndStatus(request, statusRef.get()))
-        );
+        final List<DataSegmentChangeRequestAndStatus> result = new ArrayList<>(statusRefs.size());
+        statusRefs.forEach((request, statusRef) -> {
+          // Remove complete statuses from the cache
```

Comment:
As long as the coordinator doesn't give up on a request it has sent to the historical, we should be good.
i.e. once the coordinator has asked a historical to LOAD a segment, it should not ask the historical to DROP the same segment until the LOAD has failed or succeeded.

Otherwise, we might keep the cached value of LOAD success around and send serve it to the second load request which comes after the drop.

Reply:
This code could probably do with some cleanup and doing what had been originally suggested [in this comment](https://github.com/apache/druid/pull/11717#discussion_r723833447) i.e. maintaining the cache (or really just a concurrent map) at the `DataSegment` level itself rather than `DataSegmentChangeRequest`.

But I wanted to keep the footprint of this PR low so that we could include it in Druid 27 release. We can take up a more holistic overhaul later. Let me know what you think.

Reply:
> once the coordinator has asked a historical to LOAD a segment, it should not ask the historical to DROP the same segment until the LOAD has failed or succeeded.

Could this happen after Coordinator leader change or restart?

Reply:
Yes, I feel those are the only cases when this is likely to happen.